dtm0/dtx.c

```diff
@@ -0,0 +1,276 @@
+	struct m0_dtm0_ts now;
+
+	rc = m0_dtm0_clk_src_now(&dtx->dd_dtms->dos_clk_src, &now);
+	if (rc != 0)
+		return rc;
+
+	dtx->dd_txd.dtd_id = (struct m0_dtm0_tid) {
+		.dti_ts  = now,
+		.dti_fid = dtx->dd_dtms->dos_generic.rs_service_fid,
+	};
+
+	M0_POST(m0_dtm0_tid__invariant(&dtx->dd_txd.dtd_id));
+	return rc;
+}
+
+static int m0_dtm0_dtx_open(struct m0_dtm0_dtx  *dtx,
+			    uint32_t             nr)
+{
+	return m0_dtm0_tx_desc_init(&dtx->dd_txd, nr);
+}
+
+static void m0_dtm0_dtx_assign(struct m0_dtm0_dtx  *dtx,
+			       uint32_t             pa_idx,
+			       const struct m0_fid *pa_fid)
+{
+	struct m0_dtm0_tx_pa   *pa;
+	struct m0_reqh         *reqh;
+	struct m0_conf_cache   *cache;
+	struct m0_conf_obj     *obj;
+	struct m0_conf_process *proc;
+	struct m0_fid           rdtms_fid;
+	int rc;
+
+	M0_PRE(dtx);
+	M0_PRE(pa_idx < dtx->dd_txd.dtd_pg.dtpg_nr);
+
+	/* TODO: Should we release any conf objects in the end? */
+
+	reqh = dtx->dd_dtms->dos_generic.rs_reqh;
+	cache = &m0_reqh2confc(reqh)->cc_cache;
+
+	obj = m0_conf_cache_lookup(cache, pa_fid);
+	M0_ASSERT_INFO(obj != NULL, "User service is not in the conf cache?");
+
+	obj = m0_conf_obj_grandparent(obj);
+	M0_ASSERT_INFO(obj != NULL, "Process the service belongs to "
+		       "is not a part of the conf cache?");
+
+	proc = M0_CONF_CAST(obj, m0_conf_process);
+	M0_ASSERT_INFO(proc != NULL, "The grandparent is not a process?");
+
+	rc = m0_conf_process2service_get(&reqh->rh_rconfc.rc_confc,
+					 &proc->pc_obj.co_id, M0_CST_DTM0,
+					 &rdtms_fid);
+
+	M0_ASSERT_INFO(rc == 0, "Cannot find remote DTM service on the remote "
+		       "process that runs this user service?");
+
+	pa = &dtx->dd_txd.dtd_pg.dtpg_pa[pa_idx];
+	M0_PRE(M0_IS0(pa));
+
+	pa->pa_fid = rdtms_fid;
+	M0_ASSERT(pa->pa_state == M0_DTPS_INIT);
+	pa->pa_state = M0_DTPS_INPROGRESS;
+
+	M0_LOG(DEBUG, "pa: " FID_F " (User) => " FID_F " (DTM) ",
+	       FID_P(pa_fid), FID_P(&rdtms_fid));
+}
+
+static int m0_dtm0_dtx_close(struct m0_dtm0_dtx *dtx)
+{
+	M0_PRE(dtx);
+
+	m0_sm_state_set(&dtx->dd_sm, M0_DDS_INPROGRESS);
+
+	/* TODO: add a log entry */
+
+	return 0;
+}
+
+M0_INTERNAL int m0_dtx_dtm0_init(struct m0_dtm0_service *svc,
+				 struct m0_sm_group     *group,
+				 struct m0_dtx         **out)
+{
+	struct m0_dtm0_dtx *dtx;
+	int                 rc;
+
+	if (svc == NULL) {
+		*out = NULL;
+		return 0;
+	}
+
+	rc = m0_dtm0_dtx_init(svc, group, &dtx);
+	if (rc != 0)
+		return rc;
+
+	*out = &dtx->dd_ancient_dtx;
+	return 0;
+}
+
+M0_INTERNAL void m0_dtx_dtm0_fini(struct m0_dtx **pdtx)
+{
+	if (*pdtx == NULL) {
+		return;
+	}
+
+	m0_dtm0_dtx_fini((*pdtx)->tx_dtx);
+	*pdtx = NULL;
+}
+
+M0_INTERNAL int m0_dtx_dtm0_prepare(struct m0_dtx *dtx)
+{
+	return dtx == NULL ? 0 : m0_dtm0_dtx_prepare(dtx->tx_dtx);
+}
+
+M0_INTERNAL int m0_dtx_dtm0_open(struct m0_dtx  *dtx, uint32_t nr)
+{
+	return dtx == NULL ? 0 : m0_dtm0_dtx_open(dtx->tx_dtx, nr);
+}
+
+M0_INTERNAL void m0_dtx_dtm0_assign(struct m0_dtx       *dtx,
```

Comment:
I feel that this code doesn't compile.
function returns void and int at the same time return dtx == NULL ? 0 .

In 99% of cases it's better to review something which at least compiles.